usr/lib/common/asn1.c

```diff
@@ -3473,3 +3574,806 @@ CK_RV ber_decode_DHPublicKey(CK_BYTE *data,
+    CK_ULONG len, offset;
+    CK_BYTE version[] = { 0 };
+    CK_RV rc;
+    BerElement *ber;
+    BerValue *val;
+
+    /* Calculate storage for sequence */
+    offset = 0;
+    rc = 0;
+
+    rc |= ber_encode_INTEGER(TRUE, NULL, &len, NULL, sizeof(version));
+    offset += len;
+    if (opaque != NULL) {
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, opaque->ulValueLen, 0);
+        offset += len;
+    } else {
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, rho->ulValueLen, 0);
+        offset += len;
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, seed->ulValueLen, 0);
+        offset += len;
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, tr->ulValueLen, 0);
+        offset += len;
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, s1->ulValueLen, 0);
+        offset += len;
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, s2->ulValueLen, 0);
+        offset += len;
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, t0->ulValueLen, 0);
+        offset += len;
+        if (t1) {
+            ber = ber_alloc_t(LBER_USE_DER);
+            rc = ber_put_bitstring(ber, t1->pValue, t1->ulValueLen * 8, 0x03);
+            rc = ber_flatten(ber, &val);
+
+            ber_encode_CHOICE(TRUE, 0, NULL, &len, (CK_BYTE *)val->bv_val,
+                              val->bv_len);
+            ber_free(ber, 1);
+
+            offset += len;
+        }
+    }
+
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("ber_encode_INTEGER failed\n");
+        return CKR_FUNCTION_FAILED;
+    }
+
+    if (length_only == TRUE) {
+        rc = ber_encode_SEQUENCE(TRUE, NULL, &len, NULL, offset);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_encode_SEQUENCE failed\n");
+            return rc;
+        }
+        rc = ber_encode_PrivateKeyInfo(TRUE,
+                                       NULL, data_len,
+                                       NULL, ber_AlgIdDilithiumLen,
+                                       NULL, len);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_encode_PrivateKeyInfo failed\n");
+            return rc;
+        }
+        return rc;
+    }
+
+    /* Allocate storage for sequence */
+    buf = (CK_BYTE *) malloc(offset);
+    if (!buf) {
+        TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+        return CKR_HOST_MEMORY;
+    }
+    offset = 0;
+    rc = 0;
+
+    /* Version */
+    rc = ber_encode_INTEGER(FALSE, &buf2, &len, version, sizeof(version));
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("ber_encode_INTEGER of version failed\n");
+        goto error;
+    }
+    if (buf2 != NULL) {
+        memcpy(buf + offset, buf2, len);
+        offset += len;
+        free(buf2);
+        buf2 = NULL;
+    }
+
+    /* Check if key given as opaque blob */
+    if (opaque != NULL) {
+        // the CKA_IBM_OPAQUE attrib
+        rc = ber_encode_OCTET_STRING(FALSE, &buf2, &len,
+                                     (CK_BYTE *) opaque +
```

Comment:
you do "attr + sizeof(CK_ATTRIBUTE)" at several places in this function.  Shouldn't this be "attr->pValue" instead? "attr + sizeof(CK_ATTRIBUTE)" only works if the attribute was allocated that way, that the value is directly following the CK_ATTRIBUET structure. I would not assume that this is always the case, you never know where you get the inout attributes from.

Reply:
Good point. I changed this code to using attr->pValue, but also put on our todo list that a couple of other functions (ber_encode_RSAPrivateKey, ber_encode_RSAPublicKey, ber_encode_DSAPrivateKey, ber_encode_ECPrivateKey ...) are coded in the same way.